File: Code/ros2/src/cpp_echo/src/listener.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
using std::placeholders::_1;

class Listener : public rclcpp::Node
{
    public:
    Listener()
    : Node("listener")
    {
        publisher_ = this->create_publisher<std_msgs::msg::String>("topic_echo", 10);

        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "topic", 10, std::bind(&Listener::listener_callback, this, _1));
    }
    
    private:
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;

    void listener_callback(const std_msgs::msg::String::SharedPtr msg)
    {
        if (msg->data == "exit")
        {
            rclcpp::shutdown();
            return;
        }

        RCLCPP_INFO(this->get_logger(), "I heard: %d Bytes", msg->data.length());

        auto message = std_msgs::msg::String();
        message.data = msg->data;
        publisher_->publish(message);
        // RCLCPP_INFO(this->get_logger(), "Publishing: \"%s\"", message.data.c_str());
    }
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Listener>());
    rclcpp::shutdown();
    return 0;
}

File: Code/ros2/src/py_echo/py_echo/listener.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String

from rclpy.clock import Clock
from datetime import datetime

class Listener(Node):

    def __init__(self):
        super().__init__('listener')
        self.publisher_ = self.create_publisher(String, 'topic_echo', 10)
        self.subscription = self.create_subscription(
            String,
            'topic',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning

    def listener_callback(self, msg):
        if (msg.data == 'exit'):
            raise SystemExit

        self.get_logger().info(f"I heard: {len(msg.data)} Bytes")
        self.sender(msg)

    def sender(self, msg):
        self.publisher_.publish(msg)
        #self.get_logger().info('Publishing to topic_echo: "%s"' % msg.data)


def main(args=None):
    rclpy.init(args=args)

    listener = Listener()

    try:
        rclpy.spin(listener)
    except SystemExit:
        pass

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    listener.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
